sklearn/_build_utils/__init__.py

```diff
@@ -70,20 +71,30 @@ def cythonize_extensions(extension):
         os.environ.get("SKLEARN_ENABLE_DEBUG_CYTHON_DIRECTIVES", "0") != "0"
     )
 
+    compiler_directives = {
+        "language_level": 3,
+        "boundscheck": cython_enable_debug_directives,
+        "wraparound": False,
+        "initializedcheck": False,
+        "nonecheck": False,
+        "cdivision": True,
+    }
+
+    # TODO: once Cython 3 is released and we require Cython>=3 we should get
+    # rid of the legacy_implicit_noexcept directive. This should mostly consist
+    # in ensuring nogil is at the end of function signature, e.g. replace
+    # "nogil except -1" by "except -1 nogil". See
+    # https://github.com/cython/cython/issues/5088 for more details
```

Comment:
IIRC, we also need to `noexcept`-qualify some Cython and externalized C interfaces when we drop `legacy_implicit_noexcept`. 
```suggestion
    # rid of the `legacy_implicit_noexcept` directive.
    # This should mostly consist in:
    # 
    #   - ensuring nogil is at the end of function signature,
    #   e.g. replace "nogil except -1" by "except -1 nogil".
    # 
    #   - "noexcept"-qualifying Cython and externalized C interfaces
    #   which aren't raising nor propagating exceptions. 
    #   See: https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values  # noqa
    # 
    # See: https://github.com/cython/cython/issues/5088 for more details
```